Implementing the use of TrainingDataImporter in rasa test

Imod7 opened the pull request:
Implementing the 2nd Part of the changes described in [Issue 5986](https://github.com/RasaHQ/rasa/issues/5986) related to `run_evaluation`.
_The current PR is a continuation of this one [#7539](https://github.com/RasaHQ/rasa/pull/7539) - in the 7539 I had issues with the CLA certificate and then squash and rebase became very difficult because of merged commits so I decided to create a new one, the current one._

**Proposed changes**:
- Loading `test data` using `TrainingDataImporter` in `run_evaluation` function.
- Changed the `Optional` parameters in `load_from_dict` of TrainingDataImporter to make it work with the data we have in `run_evaluation`.
- Instead of having 1 function, the `run_nlu_test` now we have 2 functions described below : 
   1. `run_nlu_test` that runs `run_nlu_test_async` in the event loop.
   2. `run_nlu_test_async`
- Passing directly the specific arguments to `run_nlu_test_async` instead of `argparse.Namespace` thus making it more reusable.
- Made async the following functions : 
  1. `run_evaluation`
  2. `compare_nlu`
  3. `test_nlu`
  4. `compare_nlu_models`
  5. `train_nlu_async` 
  6. `test_nlu_comparison`
- Making `train_nlu_async` function public so we can call it from other modules.

wochinge commented on rasa/train.py:
```diff
@@ -697,19 +697,17 @@ def train_nlu(
-            model_to_finetune=model_to_finetune,
-            finetuning_epoch_fraction=finetuning_epoch_fraction,
-        )
+    return await _train_nlu_async(
```
why was this change required?

Imod7 replied:
Because I had errors of type `RuntimeError: this event loop is already running.`. 

I just checked it again and for example for these tests  : 
```
FAILED tests/test_train.py::test_train_nlu_temp_files - RuntimeError: this event loop is already running.
FAILED tests/test_train.py::test_train_nlu_wrong_format_error_message - RuntimeError: this event loop is alre...
FAILED tests/test_train.py::test_train_nlu_with_responses_no_domain_warns - RuntimeError: this event loop is ...
FAILED tests/test_train.py::test_train_nlu_with_responses_and_domain_no_warns - RuntimeError: this event loop...
FAILED tests/test_train.py::test_train_nlu_no_nlu_file_error_message - RuntimeError: this event loop is alrea...
FAILED tests/test_train.py::test_train_nlu_autoconfig - RuntimeError: this event loop is already running.
```
I get the above errors when I have the `run_in_loop`. When I change to `await` the tests pass.

wochinge replied:
Mhm, interesting 🤔  Why did this error happen? What's the stack trace? Could it be that [this](https://github.com/RasaHQ/rasa/blob/a7387f445604be5d123b776a2f4d1521990b8453/rasa/nlu/test.py#L1900) should rather call the `async` version of `train_nlu` (which is `_train_nlu_async`)?

Imod7 replied:
@wochinge Hey Tobias! 
Here are the error messages [trace.txt](https://github.com/RasaHQ/rasa/files/5768791/trace.txt)
I can check and try to call `_train_nlu_async`

wochinge replied:
I double checked for `test_train_nlu_temp_files` and if you don't make the test function `async`, you can use `rasa.utils.common.run_in_loop` as before. 

The issue is that when you define a test function via

```python
async def test_something():
   pass
```
it will start a event loop. When your test code starts another event loop via `rasa.utils.common.run_in_loop` you get the error above. As you don't need the `async` for anything in this test, you can remove it safely.

wochinge commented on rasa/cli/test.py:
```diff
@@ -167,7 +168,16 @@ def run_nlu_test(args: argparse.Namespace) -> None:
+        await test_nlu(model_path, nlu_data, output, vars(args))
+
+
+def run_nlu_test(args: argparse.Namespace) -> None:
```
As discussed in the other PR, I think it would be better to pass the arguments directly to `run_nlu_test_async` to make `run_nlu_test_async` more reusable. Our command-line parser library `argparse` is the only thing which encapsulates arguments into the `arparse.Namespace` type, while other code would rather pass the arguments directly.

What do you think? 

That's a short and not complete example of what I mean:

```python
def run_nlu_test(args: Argparse) -> None:
   rasa.utils.common.run_in_loop(run_nlu_test_async(args.nlu, args.out, args.config, ...))

async def run_nlu_test_async(data_path: Optional[Text], output_dir: Optional[Text], config: Optional[Text], ...) -> None:
    # implementation
```

Imod7 replied:
> @Imod7 I don't understand this part in your proposed changes, what do you mean?
> 
> > Making async a list of functions and then adding the corresponding await statements when they are called.

@wochinge For someone who doesn't know the issue at all, I just thought it would be good to mention that I added `async` and `await` statements in the definition and the calling of a list of functions. Like a summary of some of the changes I did instead of mentioning the name of the functions one by one and where those changes took place. They can also look at the files that were changed but I thought it is helpful for someone to get a quick idea before checking the files.

Imod7 replied:
> As discussed in the other PR, I think it would be better to pass the arguments directly to `run_nlu_test_async` to make `run_nlu_test_async` more reusable. Our command-line parser library `argparse` is the only thing which encapsulates arguments into the `arparse.Namespace` type, while other code would rather pass the arguments directly.
> 
> What do you think?
> 
> That's a short and not complete example of what I mean:
> 
> ```python
> def run_nlu_test(args: Argparse) -> None:
>    rasa.utils.common.run_in_loop(run_nlu_test_async(args.nlu, args.out, args.config, ...))
> 
> async def run_nlu_test_async(data_path: Optional[Text], output_dir: Optional[Text], config: Optional[Text], ...) -> None:
>     # implementation
> ```

@wochinge Yes! I wanted to ask you about this one because I didn't know how to change it, like what type of arguments to pass. So it is better to pass explicitly the arguments that I will use inside the function (data_path, output_dir, config)? Can I not use something like **kwargs? Because if I pass them explicitly and at one point we add an extra option then we have to add it in the parameters also.

wochinge replied:
> So it is better to pass explicitly the arguments that I will use inside the function (data_path, output_dir, config)?

Yes 👍🏻 This has the following advantage:
* it's clear to the caller which arguments are expected
* We can type annotate each parameter which then our CI can check which in turn makes our code more robust

> Because if I pass them explicitly and at one point we add an extra option then we have to add it in the parameters also.

I'd argue that it is the responsibility of the person adding the extra option to make sure it's used correctly. If you'd use `kwargs` then we'd still have to extra the value at some point in the code. This means the person adding the extra option would have to change other code anyway. 
Or maybe from a different perspective: If all our code just passes around some dictionaries then we end up with a giant black box. When you want to understand whether / how something is at some point a provided parameters in such a codebase you'd have to understand a lot of prior code. Does that make sense?

> didn't know how to change it, like what type of arguments to pass.

To see which parameters you need, you can look at the function and see where we call `args.<some value name>`. To see which type they should have you can look at the functions which our function calls and see how these values are annotated there. Consider the following example:

```python
def foo(args: argparse.Namespace) -> None:
   my_param = args.test
   bar(my_param)
   bar2(args.output_directory)
   
def bar(value: Text, some_value: Optional[float] = None) -> None:
   pass

def bar(dir: Optional[Text] = None) -> None:
   pass
```

By looking at the implementation of `foo` you can see that there are two CLI parameters `test` and `output_directory`. After looking at the type annotation in `bar` and `bar2` you can see that they have type `Text` and type `Optional[Text]`. The explicit version of `foo` would hence look like:

```python
def foo_explicit(my_param: Text, output_directory: Optional[Text) -> None:
   bar(my_param)
   bar2(output_directory)
```

Imod7 replied:
@wochinge Changed the parameters of `run_nlu_test_async` so it's called with the specific arguments and not with `argparse.Namespace` anymore. 
Some thoughts : 
- Maybe passing `vars(args)` as a first argument (which goes to `all_args: Optional[Dict]`) is not a good idea and I should : 
   - construct the `Dict` inside the `run_nlu_test_async` function by using all the available parameters or
   - construct the `Dict` inside the `run_nlu_test`, then passing it as a Dict argument in `run_nlu_test_async`
That way I avoid any black boxes that you mention in your comments.
- Default values are set in the parameters. But every time someone decides to change the default values in general, then someone also has to change the default values in `run_nlu_test_async`. Is this a good practice? Maybe these default values are set in some global variables somewhere? 
- I added the `docstring` from my understanding of the use of each function parameter.

wochinge replied:
> Maybe passing vars(args) as a first argument (which goes to all_args: Optional[Dict]) is not a good idea and I should 

You have a very good point there. This whole `vars` thing is in my opinion not clean at all. I'm a bit worried though that this PR is getting to big, so I'd actually keep your solution for now. If you want you can tackle this in a follow-up PR?

> Default values are set in the parameters. 

I'd omit them for now as they shouldn't be required. 

> I added the docstring from my understanding of the use of each function parameter.
 Write
 
 Sounds good. Will review them in the final review 👍🏻

wochinge reviewed (commented):
Great iteration! 💯 Left a few comments

wochinge commented:
@Imod7 I don't understand this part in your proposed changes, what do you mean?

> Making async a list of functions and then adding the corresponding await statements when they are called.

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
-def run_nlu_test(args: argparse.Namespace) -> None:
-    """Run NLU tests."""
+async def run_nlu_test_async(
+    all_args: Optional[Dict] = None,
```
I'd suggest to make this the last argument (similar as `additional_arguments` in `perform_nlu_cross_validation` or `test_nlu`

Imod7 replied:
Moved `all_args` as the last parameter in `run_nlu_test_async`.

wochinge commented on rasa/__main__.py:
```diff
@@ -113,7 +114,10 @@ def main() -> None:
             rasa.telemetry.initialize_telemetry()
             rasa.telemetry.initialize_error_reporting()
-            cmdline_arguments.func(cmdline_arguments)
+            if inspect.iscoroutinefunction(cmdline_arguments.func):
+                rasa.utils.common.run_in_loop(cmdline_arguments.func(cmdline_arguments))
+            else:
+                cmdline_arguments.func(cmdline_arguments)
```
this is no longer needed, no?

Imod7 replied:
True! I removed the coroutine check and it calls directly `cmdline_arguments.func(cmdline_arguments)` as before.

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
-def run_nlu_test(args: argparse.Namespace) -> None:
-    """Run NLU tests."""
+async def run_nlu_test_async(
+    config: Optional[Union[Text, Dict]],
```
according to your docstring it should be this, no?
```suggestion
    config: Optional[Union[Text, List[Text]],
```

Imod7 replied:
True! I changed it as you suggested.

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+    """Runs NLU tests.
+
+    Args:
+        all_args: all arguments gathered in a Dict so we can pass them to 'perform_nlu_cross_validation' and
```
can please wrap the docstring so that it doesn't exceed our 88 char limit?

Imod7 replied:
I did! I also changed the setting in Pycharm so that the wrapping line is at 88 characters and this hopefully does not happen again.

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+                  'test_nlu' as parameter.
+        config: config file or a list of multiple config files.
+        data_path: path for the nlu data.
+        models_path: the path for the nlu data.
```
```suggestion
        models_path: Path to a trained Rasa model..
```

Imod7 replied:
Thank you for the suggestion. Changed it!

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+        data_path: path for the nlu data.
+        models_path: the path for the nlu data.
+        output_dir: the directory for the results to be saved.
+        cross_validation: boolean value that indicates if it should test the model using cross validation or not.
```
I suggest not repeating the types in docstrings as we have the type annotations for this.
```suggestion
        cross_validation: indicates if it should test the model using cross validation or not.
```

Imod7 replied:
True! I removed all types from doc strings.

wochinge replied:
slightly related and nice to know: https://en.wikipedia.org/wiki/Hungarian_notation

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+        output_dir: the directory for the results to be saved.
+        cross_validation: boolean value that indicates if it should test the model using cross validation or not.
+        percentages: defines the exclusion percentage of the training data.
+        runs: used in 'compare_nlu_models' and indicates the number of comparison runs.
```
```suggestion
        runs: Number of comparison runs to make.
```

Imod7 replied:
Changed it !

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+        cross_validation: boolean value that indicates if it should test the model using cross validation or not.
+        percentages: defines the exclusion percentage of the training data.
+        runs: used in 'compare_nlu_models' and indicates the number of comparison runs.
+        no_errors: boolean value that indicates if incorrect predictions should be written to a file or not.
```
I suggest not repeating the types in docstrings as we have the type annotations for this.

I suggest not repeating the types in docstrings as we have the type annotations for this.


```suggestion
        no_errors: indicates if incorrect predictions should be written to a file or not.
```

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+    models_path: Optional[Text],
+    output_dir: Optional[Text],
+    cross_validation: Optional[bool],
+    percentages: Optional[List[int]],
```
As `compare_nlu_models` only accepts a `List[int]` this should also only be able to accept `List[int]`
```suggestion
    percentages:List[int],
```

Imod7 replied:
Changed it!

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+    data_path: Optional[Text],
+    models_path: Optional[Text],
+    output_dir: Optional[Text],
+    cross_validation: Optional[bool],
```
As the our `argparse` config specifies a default, `None` can't happen
```suggestion
    cross_validation:bool,
```

Imod7 replied:
True! Changed it!

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+    output_dir: Optional[Text],
+    cross_validation: Optional[bool],
+    percentages: Optional[List[int]],
+    runs: Optional[int],
```
As the our `argparse` config specifies a default, `None` can't happen

```suggestion
    runs: int,
```

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+    percentages: Optional[List[int]],
+    runs: Optional[int],
+    no_errors: Optional[bool],
+    all_args: Optional[Dict[Text, Any]],
```
How can this parameter become `None` in the existing calls?
```suggestion
    all_args: Dict[Text, Any],
```

Imod7 replied:
True! Changed it!

wochinge commented on rasa/cli/test.py:
```diff
@@ -148,26 +171,46 @@ def run_nlu_test(args: argparse.Namespace) -> None:
+
+
+def run_nlu_test(args: argparse.Namespace) -> None:
+    """Adding this function layer to be able to run run_nlu_test_async in the event loop.
```
can you please update the docstring?

Imod7 replied:
Maybe this one 
```
"""Runs NLU tests by calling run_nlu_test_async in the event loop.

    Args:
        args: all arguments that were given and parsed.
    """
```

wochinge replied:
The `args` part feels a bit long. How about `the parsed CLI arguments for 'rasa test nlu'`?

Imod7 replied:
I adjusted it!

wochinge commented on rasa/test.py:
```diff
@@ -110,6 +112,7 @@ def test_core(
     output: Text = DEFAULT_RESULTS_PATH,
     additional_arguments: Optional[Dict] = None,
 ) -> None:
+    """Checks if models are present and then runs the test function in the event loop."""
```
I think it's better to describe what and why things are happening in a docstring instead of describing how things are done. 
```suggestion
    """Tests a trained Core model against a set of test stories."""
```

Imod7 replied:
Changed it! I will keep this advice also for all the future docstrings.

wochinge reviewed (commented):
Wuhuu, we're nearly there 💯 Awesome work with disentangling all the `async` things 🚀

Imod7 commented:
I think I did all the last changes that you mentioned and I also removed `Optional` from these parameters in `run_nlu_test_async` :
```
data_path: Text,
models_path: Text,
output_dir: Text,
no_errors: bool,
```
because as you showed yesterday I checked the functions related to `set_test_nlu_arguments` and : 
- in `add_nlu_data_param` it adds default nlu/data path `data`
- in `add_model_param` it adds the default models path/directory `models`
- in `add_test_nlu_argument_group`  it sets the default output directory `results`
- in `add_errors_success_params` it sets default no_errors `False`

Changed the docstring again of `run_nlu_test_async`  and `run_nlu_test` (edited)

wochinge commented on rasa/__main__.py:
```diff
@@ -3,6 +3,7 @@
 import os
 import platform
 import sys
+import inspect
```
this import is unused

wochinge replied:
@Imod7 This is still here?

Imod7 replied:
Yes, I was waiting for feedback on the other 2 comments and then I would push. I will do a final check and push now.

wochinge replied:
👍🏻 Could please also resolve the comments further up in the PR which you've worked in? Makes it easier to see which changes weren't addressed yet

Imod7 replied:
Yes! I will change the description of the PR and convert it from draft also.

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,54 @@ def run_core_test(args: argparse.Namespace) -> None:
+async def run_nlu_test_async(
+    config: Optional[Union[Text, List[Text]]],
+    data_path: Text,
+    models_path: Text,
```
that can be `None`, no?
```suggestion
    models_path: Optional[Text],
```

Imod7 replied:
So I thought I should remove the `Optional` because I saw that it always gets a default value so it will never be `None` as long as it is run through the command line. 
`_StoreAction(option_strings=['-m', '--model'], dest='model', nargs=None, const=None, default='models'`

wochinge replied:
Very good point! That can indeed not be `None`. Sorry for the confusion!

wochinge commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,54 @@ def run_core_test(args: argparse.Namespace) -> None:
+    config: Optional[Union[Text, List[Text]]],
+    data_path: Text,
+    models_path: Text,
+    output_dir: Text,
```
this can also be `None`, no?
```suggestion
    output_dir: Optional[Text],
```

Imod7 replied:
Same as above

wochinge replied:
👍🏻

wochinge reviewed (commented):
Two more comments. Can you please also update the pull request description and convert your PR from a draft to a actual pull request?

wochinge commented:
Just re-request review once you're done and I have a final look + approve 🎉

Imod7 commented on rasa/cli/test.py:
```diff
@@ -115,29 +116,51 @@ def run_core_test(args: argparse.Namespace) -> None:
+    cross_validation: Optional[bool] = False,
+    percentages: Optional[List[int]] = [0, 25, 50, 75],
+    runs: Optional[int] = 3,
+    no_errors: Optional[bool] = False,
```
I removed all default values from the parameters of run_nlu_test_async

wochinge reviewed (approved):
Congrats to your first Rasa Open Source change 🚀 

Two notes:
* It's (in my opinion) ok to stay a little bit more high level in your pull request description. I can see the details in the code. In my opinion it's more important to explain why certain things where changed (e.g. why did did you have to make things `async`). No need to change anything in this PR, rather a suggestion for the future).
* Either add the label `ready-to-merge` to the PR so that the PR is automatically merged or press the green `merge` button yourself once all checks passed 🎉 (which I recommend for your first PR as pressing a button is way more fun and satisfying 😄 )
